tgg/annotations/get_ensembl_db_info.py

```diff
@@ -22,53 +24,168 @@
-            c.execute(f"SELECT {columns_str} FROM gene LEFT JOIN transcript ON canonical_transcript_id = transcript_id")
 
-            for row in c:
-                d = dict(zip(columns, row))
-                gene_id = d['transcript.stable_id']
-                transcript_id_to_dates[gene_id] = d['transcript.created_date'], d['transcript.modified_date']
+def get_gene_id_to_transcript_ids(
+        database=CURRENT_ENSEMBL_DATABASE,
+        only_protein_coding=False,
+        only_canonical_transcripts=False):
+    """Returns a dictionary mapping each Ensembl gene_id => a list of transcript ids for that gene.
+
+    Args:
+        database (str): The Ensembl database name (eg. "homo_sapiens_core_107_38")
+        only_protein_coding (bool): If True, only return protein coding genes
+        only_canonical_transcripts (bool): If True, only return canonical transcripts
+    Return:
+        dict: mapping ENSG id string to a list of ENST id strings
+    """
+
+    gene_id_to_transcript_metadata_list = get_gene_id_to_transcript_metadata(
+        database=database,
+        only_canonical_transcripts=only_canonical_transcripts,
+        only_protein_coding=only_protein_coding)
+
+    return {
+        gene_id: [
+            transcript_metadata["transcript.stable_id"] for transcript_metadata in transcript_metadata_list
+        ]
+        for gene_id, transcript_metadata_list in gene_id_to_transcript_metadata_list.items()
+    }
+
+
+def get_gene_id_to_canonical_transcript_id(database=CURRENT_ENSEMBL_DATABASE, only_protein_coding=False):
+    """Returns a dictionary mapping each Ensembl gene_id => canonical transcript id
+
+    Args:
+        database (str): The Ensembl database name (eg. "homo_sapiens_core_107_38")
+        only_protein_coding (bool): Only include protein-coding genes
+
+    Return:
+        dict: mapping ENSG id string to the canonical ENST id string
+    """
+
+    gene_id_to_transcript_metadata_list = get_gene_id_to_transcript_metadata(
+        database=database,
+        only_canonical_transcripts=True,
+        only_protein_coding=only_protein_coding)
+
+    gene_id_to_canonical_transcript_id = {}
+    for gene_id, transcript_metadata_list in gene_id_to_transcript_metadata_list.items():
+        if len(transcript_metadata_list) > 1:
+            raise Exception(f"{gene_id} has more than 1 canonical transcript")
+        if len(transcript_metadata_list) == 0:
+            raise Exception(f"{gene_id} has 0 canonical transcripts")
+
+        gene_id_to_canonical_transcript_id[gene_id] = transcript_metadata_list[0]["transcript.stable_id"]
+
+    return gene_id_to_canonical_transcript_id
+
 
-    return transcript_id_to_dates
+def get_gene_created_modified_dates(
+        database=CURRENT_ENSEMBL_DATABASE,
+        only_protein_coding=False,
+        only_canonical_transcripts=False):
+    """Returns a dictionary mapping each Ensembl gene_id => a 2-tuple containing the created date and the modified date
```

Comment:
why do you need to know when the gene was created vs modified?

Reply:
I was using this to generate histograms of gene-created times and gene-last-modified times because I was curious about the rates at which OMIM has been adding and updating gene records over the past ~10 years.